Src/Workspaces/Core/Portable/Utilities/TaskExtensions.cs

```diff
@@ -337,5 +338,137 @@ public static T WaitAndGetResult<T>(this Task<T> task, CancellationToken cancell
                     _ => continuationFunction(t), cancellationToken, TaskContinuationOptions.None, scheduler),
                 cancellationToken, taskContinuationOptions, scheduler).Unwrap();
         }
+
+        /// <summary>
+        /// Wraps a task with one that will complete as cancelled based on a cancellation token, 
+        /// allowing someone to await a task but be able to break out early by cancelling the token.
+        /// </summary>
+        /// <typeparam name="T">The type of value returned by the task.</typeparam>
+        /// <param name="task">The task to wrap.</param>
+        /// <param name="cancellationToken">The token that can be canceled to break out of the await.</param>
+        /// <returns>The wrapping task.</returns>
+        public static Task<T> WithCancellation<T>(this Task<T> task, CancellationToken cancellationToken)
```

Comment:
For Roslyn tasks, we expect cancellation to be responded to quickly. Since we're a system that cancels tasks very heavily -- a simple keystroke can cancel dozens of tasks being used to compute all sorts of background work -- letting that work carry over isn't acceptable. That said, you do raise a good question what we do for analyzers when those run longer than we might like. @heejaechang, how does that work?